modo/helpers.py

```diff
@@ -23,13 +24,79 @@ def dict_to_instance(element: Mapping[str, Any]) -> Any:
     )
 
 
+def set_part_of_relationship(
+    element: model.DataEntity
+    | model.Sample
+    | model.Assay
+    | model.ReferenceGenome
+    | model.ReferenceSequence,
+    element_path: str,
+    partof_group: zarr.hierarchy.Group,
+):
+    """Add element to the hasPart attribute of a parent zarr group"""
+    parent_type = getattr(
+        model,
+        partof_group.attrs.get("@type"),
+    )
+
+    has_prop = get_haspart_property(element.__class__.__name__)
+    parent_slots = parent_type.__match_args__
+    if has_prop not in parent_slots:
+        raise ValueError(
+            f"Cannot make {element.id} part of {partof_group.name}: {parent_type} does not have property {has_prop}"
+        )
+    # has_part is multivalued
+    if has_prop not in partof_group.attrs:
+        partof_group.attrs[has_prop] = []
+    partof_group.attrs[has_prop] += [element_path]
+
+
+class UserElementType(str, Enum):
+    """Enumeration of element types exposed to the user."""
+
+    SAMPLE = "sample"
+    ASSAY = "assay"
+    DATA_ENTITY = "data"
+    REFERENCE_GENOME = "reference"
+
+    def get_target_class(
+        self,
+    ) -> type:
+        """Return the target class for the element type."""
+        if self == UserElementType.SAMPLE:
+            return model.Sample
+        elif self == UserElementType.ASSAY:
+            return model.Assay
+        elif self == UserElementType.DATA_ENTITY:
+            return model.DataEntity
+        elif self == UserElementType.REFERENCE_GENOME:
+            return model.ReferenceGenome
+        else:
+            raise ValueError(f"Unknown element type: {self}")
+
+    @classmethod
+    def from_object(cls, obj):
+        """Return the element type from an object."""
+        if isinstance(obj, model.Sample):
+            return UserElementType.SAMPLE
+        elif isinstance(obj, model.Assay):
+            return UserElementType.ASSAY
+        elif isinstance(obj, model.DataEntity):
+            return UserElementType.DATA_ENTITY
+        elif isinstance(obj, model.ReferenceGenome):
+            return UserElementType.REFERENCE_GENOME
+        else:
+            raise ValueError(f"Unknown object type: {type(obj)}")
+
+
 class ElementType(str, Enum):
-    """Enumeration of element types."""
+    """Enumeration of all element types."""
 
     SAMPLE = "sample"
     ASSAY = "assay"
     DATA_ENTITY = "data"
     REFERENCE_GENOME = "reference"
+    REFERENCE_SEQUENCE = "reference"
```

Comment:
**note:** I think it could cause issues down the line if two variants have the same value.
**suggestion:** maybe something like `sequence`, or `reference_sequence`?

Reply:
yes, that is a fair point. I was a bit unsure about this. How do we do with the `Elementtype.from_object()` function? 
Do we dosomething like:
```py
@classmethod
    def from_object(cls, obj):
        """Return the element type from an object."""
        if isinstance(obj, model.Sample):
            return ElementType.SAMPLE
       ...
        elif isinstance(obj, model.ReferenceSequence):
            return "reference"
        else:
            raise ValueError(f"Unknown object type: {type(obj)}")
```
Or is there a more elegant way to solve this? We could also go with a modo structure including an "own" group for sequence like this? :
```bash
/
 ├── assay
 │   └── assay1
 ├── data
 │   └── demo1
 ├── reference
 │   └── reference1
 ├── sequence
 │   └── seq1
 └── sample
     └── sample1
```

Reply:
I chose the second approach in https://github.com/sdsc-ordes/modo-api/pull/17/commits/7f6bcad1462b4933e534baa570fe562a4ea46030 with a separate `sequence` zarr group, because it allows a coherent use of `ElementType` without adding extra logic for the reference sequences. 
Let me know if you think we should only expose the groups `UserElementTypes` on the zarr tree and add ReferenceSequences to the reference group instead? I'm really not sure about this and see advantages for both sides. Adaptations were minor and we can easily rollback or readjust? 

Other adaptations in https://github.com/sdsc-ordes/modo-api/pull/17/commits/62e476f89ee2bfa4d01c1be36ab853f7fe8b435a and https://github.com/sdsc-ordes/modo-api/pull/17/commits/315ec9ea777bf3f175b45604bfb03c4299ce03bc are related to adjusting ids in `has_part` relationships to full ids when building modo from yaml file. I realized that they were still using the short ids (`element_name`) instead of full ids (`element_type / element_name`).

Reply:
Nice!! That makes sense :smiley: